### PR TITLE
[#19510][feat] Add native Qwen3 reranking support

### PR DESCRIPTION
**Pull request by anurags25**
## Description

Adds native Qwen3-Reranker support to the PyTorch backend.

- Adds `Qwen3ForTextReranking`, deriving a scalar `yes - no` scoring head from tied or untied checkpoint weights.
- Uses the encode-only executor with dynamic batching.
- Adds `trtllm-serve rerank` with `/rerank`, `/v1/rerank`, and `/v2/rerank`.
- Implements Qwen3 prompt formatting, document truncation, scoring, and API documentation.
- Supports the Qwen3-Reranker 0.6B, 4B, and 8B checkpoints.

Closes #19510.

## Test Coverage

- 23 targeted unit tests passed.
- Hugging Face logit parity passed for the Qwen3 reranker 0.6B and 8B checkpoints on H200.
- Live HTTP validation covered all three endpoints, truncation, `top_n`, error handling, and concurrent requests.

## Dev Engineer Review

- Adds native `Qwen3ForTextReranking` support with a scalar `yes_logit - no_logit` score head. Supports tied and untied weights and avoids generation-time vocabulary projection.
- Adds encode-only dynamic batching through `trtllm-serve rerank`.
- Adds `/rerank`, `/v1/rerank`, and Cohere-compatible `/v2/rerank` APIs.
- Implements Qwen3 prompt formatting, document truncation, sigmoid score conversion, `top_n`, validation, queue limits, and optional document responses.
- Supports Qwen3-Reranker 0.6B, 4B, and 8B routing and documentation.
- Shared encode-only configuration now rejects tensor, pipeline, or context parallelism greater than one. This constraint and the local-checkpoint architecture remapping require deployment verification.
- No current review severity findings or executed test results were supplied beyond the reported validation.

## QA Engineer Review

- Added model, routing, prompt, API, and remote-server tests. Coverage includes tied and untied weights, architecture remapping, truncation, validation errors, response formats, priorities, `top_n`, concurrency, and Hugging Face score parity.
- Updated A10 and L40S pre-merge test lists. L40S includes Qwen3 accuracy and OpenAI rerank API tests. A10 includes model, routing, and reranker tests.
- Coverage verdict: **sufficient**, subject to execution in the listed CI environments.

## Per-File QA Perspective

- `docs/source/commands/trtllm-serve/trtllm-serve.rst`: Documents the new rerank command and routes.
- `docs/source/features/reranking.md`: Documents request formats, limits, scoring, errors, batching, and supported models.
- `docs/source/index.rst`: Adds reranking documentation to the feature index.
- `docs/source/models/supported-models.md`: Documents `Qwen3ForTextReranking` support.
- `tensorrt_llm/_torch/model_config.py`: Treats reranking as non-generation behavior.
- `tensorrt_llm/_torch/models/_arch_index.py`: Routes `Qwen3ForTextReranking` to `modeling_qwen3`.
- `tensorrt_llm/_torch/models/modeling_qwen3.py`: Adds scalar reranker inference and checkpoint weight handling. Verify tied weights, untied weights, and missing-weight errors.
- `tensorrt_llm/commands/serve.py`: Adds CLI registration, architecture remapping, encode-only configuration, and rerank server startup.
- `tensorrt_llm/llmapi/disagg_utils.py`: Adds the public `ServerRole.RERANK` value.
- `tensorrt_llm/serve/encode_batcher.py`: Generalizes queue handling for embedding and rerank requests. Verify shared lifecycle and queue-full behavior.
- `tensorrt_llm/serve/openai_protocol.py`: Adds validation and response schemas for the three rerank API variants.
- `tensorrt_llm/serve/openai_server.py`: Adds rerank routes, batching, score conversion, sorting, limits, and error responses.
- `tensorrt_llm/serve/reranker.py`: Adds Qwen3 prompt construction, token truncation, suffix preservation, and stable sigmoid scoring.
- `tensorrt_llm/usage/architecture_allowlist.py`: Allows telemetry for the new public architecture.
- `tests/integration/defs/accuracy/test_llm_api_pytorch_encode.py`: Adds Hugging Face parity tests for Qwen3-Reranker 0.6B and 8B. The tests are listed in the L40S CI test list.
- `tests/integration/test_lists/test-db/l0_a10.yml`: Adds A10 CI entries for reranker modeling, routing, and tests.
- `tests/integration/test_lists/test-db/l0_l40s.yml`: Adds L40S CI entries for Qwen3 accuracy and rerank API coverage.
- `tests/unittest/_torch/modeling/test_modeling_qwen3_reranker.py`: Covers registration, generation classification, score-head derivation, lazy weight loading, missing weights, and probability conversion. The test is listed in the A10 CI list.
- `tests/unittest/llmapi/apps/_test_openai_rerank.py`: Covers `/rerank`, `/v1/rerank`, `/v2/rerank`, sorting, documents, instructions, truncation, validation, and concurrent requests. The API test is listed in the L40S CI list.
- `tests/unittest/llmapi/apps/openai_server.py`: Adds the remote rerank server fixture used by API tests.
- `tests/unittest/llmapi/test_rerank_arch_routing.py`: Covers architecture remapping, override precedence, unknown architectures, and parallelism rejection. The tests are listed in the A10 CI list.
- `tests/unittest/llmapi/test_reranker.py`: Covers prompt formatting, truncation, schema validation, response formats, limits, priorities, and invalid outputs. The tests are listed in the A10 CI list.

### CODE REVIEW

**Review by Mgluhovskoi — APPROVED** [review]

**Review by Mgluhovskoi — APPROVED** [review]

File: docs/source/commands/trtllm-serve/trtllm-serve.rst
Comment (BowenFu) [inline docs/source/commands/trtllm-serve/trtllm-serve.rst]: Thanks for putting this together! The 1.3 release branch is about to be cut from main (planned for 9/29), and until then we're asking larger changes and refactors like this one to wait so main stays stable for the branch-out. If this really needs to make 1.3 or a nightly/weekly release, could you explain why in the PR description and in your review request? I'll resolve this thread once that's there, or after branch-out either way. This is only a timing hold, not a concern with the change itself.

File: tests/integration/test_lists/test-db/l0_l40s.yml [hunk 1]
```diff
@@ -28,6 +28,10 @@ l0_l40s:
   # Qwen3-Embedding text-embedding accuracy vs HuggingFace (0.6B small/fast + 8B large)
   - accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_qwen3_text_embedding_matches_huggingface[qwen3-embedding-0.6b]
   - accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_qwen3_text_embedding_matches_huggingface[qwen3-embedding-8b]
+  # Qwen3-Reranker model accuracy and HTTP API coverage
+  - accuracy/test_llm_api_pytorch_encode.py::TestEncoderEncode::test_qwen3_reranker_matches_huggingface[qwen3-reranker-0.6b]
```
Comment (xinhe-nv) [inline tests/integration/test_lists/test-db/l0_l40s.yml]: could you add accuracy tests into qa test list too?

File: tensorrt_llm/commands/serve.py [hunk 2]
```diff
@@ -817,6 +827,40 @@ def _resolve_embedding_architecture_override(
...
+    """Route a published Qwen3-Reranker checkpoint to its scoring model."""
+    try:
+        from transformers import AutoConfig
+        hf_config = AutoConfig.from_pretrained(
+            model, trust_remote_code=trust_remote_code, revision=revision)
+        architectures = getattr(hf_config, "architectures", None) or []
+    except (OSError, ValueError) as e:
+        logger.warning(
+            f"Could not read model config for rerank routing ({model}): {e}")
+        return None
```
[the remark this answers is not in the document]
Comment (anurags25) [reply tensorrt_llm/commands/serve.py]: Fixed.

File: tests/unittest/llmapi/apps/_test_openai_rerank.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def _post(server, path, **overrides):
+    request = {
+        "model": MODEL,
+        "query": QUERY,
+        "documents": DOCUMENTS,
+    }
+    request.update(overrides)
+    return requests.post(server.url_for(*path.split("/")), json=request)
```
[the remark this answers is not in the document]
Comment (anurags25) [reply tests/unittest/llmapi/apps/_test_openai_rerank.py]: Fixed.

File: tests/unittest/llmapi/apps/_test_openai_rerank.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+def test_custom_instruction_and_document_truncation(server):
+    response = _post(
+        server,
+        "v1/rerank",
+        instruction="Judge whether the document is about a country's capital",
+        max_tokens_per_doc=8,
+    )
+    assert response.status_code == 200
+    assert len(response.json()["results"]) == len(DOCUMENTS)
```
[the remark this answers is not in the document]
Comment (anurags25) [reply tests/unittest/llmapi/apps/_test_openai_rerank.py]: Fixed.

File: docs/source/commands/trtllm-serve/trtllm-serve.rst
Comment (anurags25) [reply docs/source/commands/trtllm-serve/trtllm-serve.rst]: Hey Bowen! Yes, this can wait after the 1.3 release is cut.

File: tensorrt_llm/serve/openai_server.py [hunk 5]
```diff
@@ -1658,6 +1665,113 @@ async def openai_embedding(self, request: EmbeddingRequest,
...
+                        functools.partial(
+                            build_qwen3_rerank_input,
+                            tokenizer=self.tokenizer,
+                            query=request.query,
+                            document=document,
+                            max_seq_len=engine.max_seq_len,
+                            instruction=request.instruction,
+                            max_tokens_per_doc=request.max_tokens_per_doc))
+                    for document in documents
+                ])
```
[the remark this answers is not in the document]
Comment (anurags25) [reply tensorrt_llm/serve/openai_server.py]: Fixed.
